### PR TITLE
Add the AAQ algorithm for the ONNX backend

### PR DESCRIPTION
**Pull request by andrey-churkin**
### Changes

- Support graph with quantize-dequantize in the `find_quantizer_nodes_to_cut()` method
- Implement the CommandCreator interface for the ONNX backend
- Add the AAQ algorithm for the ONNX backend

### Reason for changes

See the ticket

### Related tickets

Ref: 127606

### Tests

tests/common/quantization/test_quantizer_removal.py

### CODE REVIEW

File: nncf/common/quantization/quantizer_removal.py [hunk 1]
```diff
@@ -62,6 +62,17 @@ def _parse_node_relatives(node: NNCFNode, is_parents: bool):
                     if relative not in to_cut:
                         to_cut.append(relative)
                     to_see_children.append(relative)
+                    # We should see parents for the `relative` node here only if they are
```
Comment (kshpv) [inline nncf/common/quantization/quantizer_removal.py:65]: Do you plan to add a test on this functionality for ONNX?

File: nncf/onnx/graph/model_transformer.py [hunk 2]
```diff
@@ -377,6 +383,24 @@ def _apply_bias_correction_transformations(
...
+    @staticmethod
+    def _apply_weight_update_transformations(
+        model: onnx.ModelProto, transformations: List[ONNXWeightUpdateCommand]
+    ) -> onnx.ModelProto:
+        """
+        :param model:
+        :param transformations:
+        :return:
+        """
+        name_to_node_map = get_name_to_node_map(model)
```
Comment (kshpv) [inline nncf/onnx/graph/model_transformer.py:395]: Please, add a test for a new transformation

File: nncf/onnx/graph/model_transformer.py [hunk 3]
```diff
@@ -99,21 +101,25 @@ def transform(self, transformation_layout: TransformationLayout) -> onnx.ModelPr
                 quantizer_insert_transformations.append(transformation)
             elif isinstance(transformation, ONNXOutputInsertionCommand):
                 output_insert_transformations.append(transformation)
-            elif isinstance(transformation, ONNXBiasCorrectionCommand):
```
Comment (kshpv) [inline nncf/onnx/graph/model_transformer.py:102]: What is the reason to move it?

**Review by kshpv — CHANGES_REQUESTED** [review]

File: nncf/common/quantization/quantizer_removal.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/common/quantization/quantizer_removal.py:65]: This test is already present. See tests/common/quantization/test_quantizer_removal.py

File: nncf/onnx/graph/model_transformer.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/onnx/graph/model_transformer.py:368]: Please take another look. The code has been reorganized, and there is no new transformation.

File: nncf/onnx/graph/model_transformer.py (hunk 3 above)
Comment (andrey-churkin) [reply nncf/onnx/graph/model_transformer.py:102]: We should update the weight after removing Q-DQ. Aligned with OV backend.

File: nncf/onnx/graph/transformations/commands.py [hunk 4]
```diff
@@ -72,18 +72,18 @@ class ONNXOutputInsertionCommand(ONNXInsertionCommand):
...
+    Update initializer in the value.
     """
 
-    def __init__(self, target_point: ONNXTargetPoint, bias_value: np.ndarray):
+    def __init__(self, target_point: ONNXTargetPoint, new_value: np.ndarray):
         """
-        :param target_point: The TargetPoint instance for the correction that contains layer's information.
-        :param bias_value: The bias shift value (numpy format) that will be added to the original bias value.
+        :param target_point: Target point.
+        :param weight_value: New weight value.
```
Comment (kshpv) [inline nncf/onnx/graph/transformations/commands.py:83]:
```suggestion
        :param new_value: New value put in the initializer.
```

File: nncf/onnx/quantization/quantize_model.py [hunk 5]
```diff
@@ -75,3 +82,91 @@ def quantize_impl(
...
+
+def quantize_with_accuracy_control_impl(
+    model: onnx.ModelProto,
+    calibration_dataset: Dataset,
+    validation_dataset: Dataset,
+    validation_fn: Callable[[Any, Iterable[Any]], Tuple[float, Union[None, List[float], List[List[TTensor]]]]],
+    max_drop: float = 0.01,
+    drop_type: DropType = DropType.ABSOLUTE,
+    preset: Optional[QuantizationPreset] = None,
+    target_device: TargetDevice = TargetDevice.ANY,
```
Comment (kshpv) [inline nncf/onnx/quantization/quantize_model.py:95]: Can we duplicate warning messages and excpetion raisers from `quantize_impl(...)`?
I would put them in the single function

File: nncf/onnx/quantization/quantize_model.py [hunk 6]
```diff
@@ -75,3 +82,91 @@ def quantize_impl(
     quantized_model = quantization_algorithm.apply(model, graph, dataset=calibration_dataset)
 
     return quantized_model
+
+
+def quantize_with_accuracy_control_impl(
```
Comment (kshpv) [inline nncf/onnx/quantization/quantize_model.py:87]: Waiting for a check of this method on yolov8.
Also, I would like to have a test on that method, probably mocking validation methods and testing on output model graph or output ignored scope....

File: nncf/quantization/algorithms/accuracy_control/onnx_backend.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        weight_name = node_with_weight.layer_attributes.weight_attrs[port_id]["name"]
+        return get_tensor_value(model, weight_name)
+
+    @staticmethod
+    def get_weight_tensor_port_ids(node: NNCFNode) -> List[Optional[int]]:
+        return list(node.layer_attributes.weight_attrs.keys())
+
+    @staticmethod
+    def get_model_size(model: onnx.ModelProto) -> int:
+        raise NotImplementedError
```
Comment (kshpv) [inline nncf/quantization/algorithms/accuracy_control/onnx_backend.py:92]: Why is this not implemented?

**Review by kshpv — CHANGES_REQUESTED** [review]

**Comment by andrey-churkin** [comment]
I've added an example for the YOLOv8 model based on the same example for the OpenVINO backend. Some comparisons are presented below to confirm the correctness of the implementation for the ONNX backend.

|  |OpenVINO|ONNX|
|--|--|--|
|Total execution time|00:04:38|00:10:11|
|Operations which were reverted to FP precision*| /model.0/conv/Conv/WithoutBiases|/model.0/conv/Conv|
|Final metric drop|0.00786| 0.00751|
|Performance speed up (throughput mode)|1.979|1.956|

*Same operations in the graph

File: nncf/onnx/quantization/quantize_model.py (hunk 6 above)
Comment (andrey-churkin) [reply nncf/onnx/quantization/quantize_model.py:88]: Results are presented below.

File: nncf/onnx/graph/transformations/commands.py (hunk 4 above)
Comment (andrey-churkin) [reply nncf/onnx/graph/transformations/commands.py]: Fixed.

File: nncf/quantization/algorithms/accuracy_control/onnx_backend.py (hunk 7 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/onnx_backend.py]: Done

**Comment by kshpv** [comment]
There is one open question - what backend should be used for ONNX model validation.
I would suggest to switch to OpenVINO validation as it was done before the same for mobilenet-v2. What do you think?

File: nncf/onnx/graph/model_transformer.py [hunk 8]
```diff
@@ -411,24 +409,51 @@ def _apply_qdq_node_removing_transformations(
...
+                model.graph.initializer.remove(initializers[quantize_node_proto.input[i]])
+                model.graph.value_info.remove(value_infos[quantize_node_proto.input[i]])
+
+            for node_proto in [quantize_node_proto, dequantize_node_proto]:
+                model.graph.value_info.remove(value_infos[node_proto.output[0]])
+
+            model.graph.node.remove(quantize_node_proto)
+            model.graph.node.remove(dequantize_node_proto)
+
+        onnx.checker.check_model(model)
```
Comment (kshpv) [inline nncf/onnx/graph/model_transformer.py:457]: Why do we have this check?

File: examples/post_training_quantization/onnx/yolov8_quantize_with_accuracy_control/main.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+from ultralytics.utils import ops
+from ultralytics.utils.metrics import ConfusionMatrix
+
+import nncf
+
+ROOT = Path(__file__).parent.resolve()
+
+
+def validate(
+    model: onnx.ModelProto, data_loader: torch.utils.data.DataLoader, validator: Validator, num_samples: int = None
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply examples/post_training_quantization/onnx/yolov8_quantize_with_accuracy_control/main.py]: Fixed

**Comment by andrey-churkin** [comment]
> There is one open question - what backend should be used for ONNX model validation. I would suggest to switch to OpenVINO validation as it was done before the same for mobilenet-v2. What do you think?

Do you mean I should use the OpenVINOExecutionProvider or convert the model to OV format? The idea was to use the ONNX model for validation, as we discussed with @alexsu52.

File: nncf/onnx/graph/model_transformer.py (hunk 8 above)
Comment (andrey-churkin) [reply nncf/onnx/graph/model_transformer.py:457]: This check failed prior to my changes. So we had an invalid ONNX proto after QDQ removing. I added this check here to be sure that ONNX proto we get her is legal.

**Comment by kshpv** [comment]
> > There is one open question - what backend should be used for ONNX model validation. I would suggest to switch to OpenVINO validation as it was done before the same for mobilenet-v2. What do you think?
> 
> Do you mean I should use the OpenVINOExecutionProvider or convert the model to OV format? The idea was to use the ONNX model for validation, as we discussed with @alexsu52.

To use pure OpenVINO for validation as well not OpenVINOExecutionProvider without conversation to OV format.The same was done for mobilenet-v2 sample

File: nncf/onnx/graph/model_transformer.py (hunk 8 above)
Comment (kshpv) [reply nncf/onnx/graph/model_transformer.py:457]: Why can not we cover this via unit tests?
Could you add a testcase for the previous failed scenario?

Comment (andrey-churkin) [reply nncf/onnx/graph/model_transformer.py:457]: Could you please explain your concern about this check? It is not clear for me.

> Could you add a testcase for the previous failed scenario?

We have such tests
https://github.com/openvinotoolkit/nncf/blob/c0e51f8833e04697da944745f7488fc4b553f832/tests/onnx/test_model_transformer.py#L197

Comment (andrey-churkin) [reply nncf/onnx/graph/model_transformer.py]: Done

**Comment by andrey-churkin** [comment]
New results after moving to the OpenVINO for validation are presented below
|  |ONNX|
|--|--|
|Total execution time|00:05:15|
|Operations which were reverted to FP precision |/model.22/Add_11|
|FP32 model metric mAP50-95(B) |0.449648|
|FP32 model metric mAP50-95(M) |0.365132|
|Final metric drop mAP50-95(B) |0.002127|
|Final metric drop mAP50-95(M) |0.000739|
|Performance speed up (throughput mode)|2.192|

**Review by kshpv — APPROVED** [review]

**Comment by andrey-churkin** [comment]
@alexsu52 Please review

File: nncf/onnx/quantization/quantize_model.py [hunk 10]
```diff
@@ -76,3 +83,91 @@ def quantize_impl(
...
+    )
+
+    should_terminate, accuracy_drop = calculate_accuracy_drop(
+        initial_metric_results.metric_value, quantized_metric_results.metric_value, max_drop, drop_type
+    )
+
+    nncf_logger.info(f"Accuracy drop: {accuracy_drop} ({drop_type})")
+
+    if advanced_accuracy_restorer_parameters.tune_hyperparams and not should_terminate:
+        pass
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/onnx/quantization/quantize_model.py:148]: Fixed. I've checked. It works without any changes.

File: nncf/quantization/algorithms/accuracy_control/onnx_backend.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+from nncf.onnx.graph.node_utils import get_bias_value
+from nncf.onnx.graph.node_utils import is_node_with_bias
+from nncf.onnx.graph.onnx_helper import get_tensor_value
+from nncf.openvino.engine import OVCompiledModelEngine
+from nncf.openvino.graph.model_utils import model_has_state
+from nncf.quantization.algorithms.accuracy_control.backend import AccuracyControlAlgoBackend
+from nncf.quantization.algorithms.accuracy_control.backend import PreparedModel
+
+
+class ONNXPreparedModel(PreparedModel):
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/onnx_backend.py:36]: Fixed

**Review by alexsu52 — APPROVED** [review]
LGTM
